### PR TITLE
feat(ec2): allow users to specify EC2 instance IDs to monitor

### LINKED ISSUE #43: [ec2] add filter of instance IDs
As a customer:

> I would just prefer a property under .monitorEC2instances that would allow me to specify instance ID's or something along those lines. I didn't deploy the instances via an ASG, so it seems like an all or nothing approach using this construct.

### PR DESCRIPTION
**Pull request by voho**
Allows users to specify instance IDs in the EC2 monitoring and combine them with ASG property.

Fixes #43 

---

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license_

### CODE REVIEW

File: lib/monitoring/aws-ec2/EC2MetricFactory.ts [hunk 1]
```diff
@@ -1,30 +1,158 @@
...
+import { DimensionsMap, IMetric } from "aws-cdk-lib/aws-cloudwatch";
 
 import { MetricFactory, MetricStatistic } from "../../common";
 
 const EC2Namespace = "AWS/EC2";
 
-export class EC2MetricFactory {
-  protected readonly metricFactory: MetricFactory;
-  protected readonly autoScalingGroup?: IAutoScalingGroup;
+export interface IECMetricFactoryStrategy {
```
Comment (ayush987goyal) [inline lib/monitoring/aws-ec2/EC2MetricFactory.ts:8]:
```suggestion
export interface IEC2MetricFactoryStrategy {
```

**Review by ayush987goyal — COMMENTED** [review]
Overall it looks good to me! Thanks for making these changes!

1. Do you think it is worthwhile to try to test this by deploying to an actual stack in an account?
2. Do you think we need to mark this change as breaking?

**Comment by voho** [comment]
> Overall it looks good to me! Thanks for making these changes!
> 

Thanks!

> 1. Do you think it is worthwhile to try to test this by deploying to an actual stack in an account?

Agree. I have tried to first extend the snapshot tests to whatever there was before, then I made changes and made sure the tests do not change any snapshot. Finally, I added new test cases with extended behaviour. But still worthwhile to test on a live stack. Any tips on how to do this? With the internal package, this was easy, maybe I can create a burner account and use this branch of the library? Any other ideas?

> 2. Do you think we need to mark this change as breaking?

You mean because of the metric name changes? It would be correct to do so, however, it depends on how many people are relying on the metric names in their overrides. Keeping it open for now, but no problem of doing it.

**Comment by ayush987goyal** [comment]
> Agree. I have tried to first extend the snapshot tests to whatever there was before, then I made changes and made sure the tests do not change any snapshot. Finally, I added new test cases with extended behaviour. But still worthwhile to test on a live stack. Any tips on how to do this? With the internal package, this was easy, maybe I can create a burner account and use this branch of the library? Any other ideas?

I think we can create a test app (similar to internal tests package) inside an `examples` folder in the repo. We can deploy this stack locally through CLI in personal/burner account and possibly post a screenshot to the PR. Ideally this should be automated by some workflow action as described in this issue: https://github.com/cdklabs/cdk-monitoring-constructs/issues/25.

For this current PR, you can just create the app locally and not push it (for now). We can create that as part of another change. Thoughts?

> You mean because of the metric name changes? It would be correct to do so, however, it depends on how many people are relying on the metric names in their overrides. Keeping it open for now, but no problem of doing it.

Was also thinking on the metric factory actually. But it should be fine I think. @echeung-amzn thoughts on this?

**Comment by ayush987goyal** [comment]
Eventually we would also do something like what `aws-cdk` repo is doing with `integ-runner`. This would be possible if and when the CDK team decides to open source `integ-runner`.

**Review by ayush987goyal — APPROVED** [review]

**Comment by echeung-amzn** [comment]
Regarding testing, I did create a [basic sample app](https://github.com/echeung-amzn/sample-cdk-monitoring-app) that I had used to deploy to an internal test account directly, although it's a bit finnicky since `npm-link` seemingly doesn't handle peer dependencies well so I ended up copying/pasting things in `node_modules` directly.

**Review by echeung-amzn — APPROVED** [review]

**Comment by echeung-amzn** [comment]
Regarding the breaking change argument, I was actually thinking of whether we should explicitly define what we consider to be part of our "public API" that the semantic versioning revolves around. It'd seem appropriate to ensure the versioning reflects what's available as the top-level facade API, but anything more "internal", despite being exported and extensible as an escape hatch, may have looser guarantees.

**Pull request merged**